htmltools/_core.py

```diff
@@ -313,10 +323,13 @@ def get_html_string(
                 )
 
             else:
-                # If we get here, x must be a string.
                 if prev_was_add_ws:
                     html_ += "  " * indent
 
+                if isinstance(child, ReprHtml):
+                    child = HTML(child._repr_html_())  # type: ignore
+
+                # At this point, child must be a string/HTML.
```

Comment:
This is added inside of the `else` code block, but I think it would be clearer if it were added at the same level as the `else`. Something like this makes it really clear what's going on (and the reader doesn't have to look at `_normalize_text` to find out that it's a no-op for `HTML` objects):

```py
            elif isinstance(child, ReprHtml):
                if prev_was_add_ws:
                    html_ += "  " * indent

                html_ += HTML(child._repr_html_())  # type: ignore[reportPrivateUsage]

                prev_was_add_ws = False
```

One more thing I wonder about: the code as currently written assumes that the `_repr_html_` method does _not_ want to have space around it. But in practice, I think that objects with `_repr_html_` are always block elements, because they're meant to be displayed beneath a notebook cell. So perhaps the whitespace logic should be changed to reflect that? Or are there any cases of `_repr_html_` where they want to be inline?

Reply:
Done in [fb78cda](https://github.com/posit-dev/py-htmltools/pull/74/commits/fb78cdac95e4d26f0ecd2a7a19c6998cb4230830). I agree that most if not all `_repr_html_` cases should be block elements, but I don't think it's terrible to say `x._repr_html()` gets the same treatment as `HTML()`, especially considering that adding whitespace correctly would add considerable complexity here without much gain (we can't control whitespace _within_ `x._repr_html()`, so producing readable HTML is mostly up to `x`).

Reply:
I just mean that if we treat it as block element, we'd do this:

```py
            elif isinstance(child, ReprHtml):
                html_ += "  " * indent

                html_ += HTML(child._repr_html_())  # type: ignore[reportPrivateUsage]

                prev_was_add_ws = True
```

Reply:
I think what you're wanting would be more involved to implement correctly, though. With your suggestion, there is nothing to handle to `eol` whitespace (here's a minimal example):

```python
import htmltools as h

class Foo:
    def _repr_html_(self) -> str:
        return "<span>Foo</span>"

str(h.span(f))
```

```html
<span>  <span>Foo</span></span>
```

As I said before, it doesn't feel worth the added complexity for such a minor detail. Plus, I like that the behavior is consistent with `HTML()`

Reply:
The behavior in your example (using the code I suggested) is the same as the current behavior if you put a block (`_add_ws=True`) element inside of an inline (`_add_ws=False`) element.

```py
import htmltools as h
print(str(h.span(h.div())))
```

```html
<span>  <div></div></span>
```

Do you think we should expect some different behavior here?

Reply:
Oh yea, I do find that surprising, but I don't think it's a big problem since that's not valid HTML